File: src/test/java/com/ottamotta/yunar/PonyServiceIntegrationTests.java
package com.ottamotta.yunar;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;

import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@RunWith(SpringRunner.class)
@SpringBootTest
public class PonyServiceIntegrationTests {

    @Autowired
    PonyService ponyService;

    @Test
    public void ponyBookedOnceWithConcurrentAccess() {


        String name = "Rainbow Dash";

        List<CompletableFuture<Pony>> bookResults =
                IntStream.range(0, 100)
                        .parallel()
                        .mapToObj(i -> CompletableFuture.supplyAsync(() -> {
                            try {
                                return ponyService.bookPony(name);
                            } catch (Exception e) {
                                return null;
                            }
                        }))
                        .collect(Collectors.toList());

        List<Pony> bookedPonies = bookResults
                .stream()
                .map(CompletableFuture::join)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());

        Assert.assertEquals("1 booking expected, but " + bookedPonies.size() + " occured", 1, bookedPonies.size());

    }


}

File: src/test/kotlin/com/ottamotta/yunar/PonyServiceIntegrationTests.kt
package com.ottamotta.yunar

import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit4.SpringRunner
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.stream.Collectors
import java.util.stream.IntStream

@RunWith(SpringRunner::class)
@SpringBootTest
class PonyServiceIntegrationTests {

    private val name = "Rainbow Dash"

    @Autowired
    lateinit var ponyService: PonyService

    @Test
    fun `pony booked only once with concurrent access`() {

        val bookResults = IntStream.range(0, 100)
                .parallel()
                .mapToObj { CompletableFuture.supplyAsync<Pony> {
                        try {
                            ponyService.bookPony(name)
                        } catch (e: Exception) {
                            null
                        }
                    }
                }
                .collect(Collectors.toList())

        val bookedPonies = bookResults
                .stream()
                .map { it.join() }
                .filter{ Objects.nonNull(it) }
                .collect(Collectors.toList())

        Assert.assertEquals("1 booking expected, but " + bookedPonies.size + " occured", 1, bookedPonies.size.toLong())

    }

    @After
    @Before
    fun restorePonyState() {
        ponyService.resetPony(name)
    }


}
